File: <100/037/037.jl
using Primes

ord(n:: Int):: Int = n |> string |> length

function left_truncatable(n:: Int):: Bool
  if n == 0
    return true
  end
  return n |> isprime ? left_truncatable(n%(10^(ord(n)-1))) : false
end

function right_truncatable(n:: Int):: Bool
  if n == 0
    return true
  end
  return n |> isprime ? right_truncatable(n÷10) : false
end

is_truncatable(n:: Int):: Bool = n |> right_truncatable ? n |> left_truncatable : false


function main():: Nothing
  upper_limit:: Int = 10^7
  count:: Int = 0
  answer:: Int = 0
  for n ∈ 10:upper_limit
    if n |> is_truncatable
      count += 1
      answer += n
    end
    if count == 11
      "11 both-truncatable primes have been found" |> println
      break
    end
  end

  if count < 11
    "only " * string(count) * " both-truncatable primes have been found below " * string(upper_limit) |> println
  end

  answer |> println
  return nothing
end

@time main()

File: <100/037/037.cpp
#include<iostream>
using namespace std;
#define UPPER_LIMIT 1000000

bool isPrime[UPPER_LIMIT];


void Eratosthenes(){
  for(int i = 2; i < UPPER_LIMIT; ++i) isPrime[i] = true;

  for(int i = 2; i < UPPER_LIMIT; ++i)if(isPrime[i]){
    for(int j = 2*i; j < UPPER_LIMIT; j += i){
      isPrime[j] = false;
    }
  }
}


int naive_pow(int n){
  int r = 1;
  while(n > 0){
    --n;
    r *= 10;
  }
  return r;
}


int clip_left(int n){
  string ns   = to_string(n);
  int ord     = ns.size() - 1;
  int divisor = naive_pow(ord);
  return n % divisor; 
}


int clip_right(int n){
  return n / 10;
}


bool truncatable(int n, int (*direc)(int)){
  while(n > 0){
    if(!isPrime[n]) return false;
    n = direc(n);
  }

  return true;
}


int main(){
  Eratosthenes();
  int answer = 0, target = 11, cnt = 0;
  for(int i = 11; i < UPPER_LIMIT; ++i){
    if(truncatable(i, clip_right) && truncatable(i, clip_left)){
      answer += i;
      ++cnt;
      if(cnt == target){
        cout << answer << endl;
        return 0;
      }
    }
  }

  cout << "only " << cnt << " both-truncatable primes have been found" << endl;
  return 0;
}
